### ISSUE TITLE
No support for log targets?

### ISSUE THREAD

**Issue by yberreby** [daboross/fern#6]
The log crate allows the use of targets, which is useful to add information about the origin of a log message, but it seems `fern` provides no facility for reading targets from `DispatchConfig`. Could this be added?

https://doc.rust-lang.org/log/log/index.html

> A log request consists of a target, a level, and a body. A target is a string which defaults to the module path of the location of the log request, though that default may be overridden. Logger implementations typically use the target to filter requests based on some user configuration.

An example from this page:
`info!(target: "yak_events", "Commencing yak shaving for {:?}", yak);`

**Comment by daboross** [comment]
Ah, yes, it seems that I've completely missed this part of the log API.

Do you think that allowing a DispatchConfig to specify a list of targets which it will run for would be sufficient to use this?

**Comment by filsmick** [comment]
I think this would be an improvement, but I think the target of a log message should also be passed as to the `format` closure of `DispatchConfig`, to allow logging it like `env_logger` does (`env_logger` prints the name of the target, which is very useful). 

Here's how this could work:

```rust
format: Box::new(|msg: &str, level: &log::LogLevel, target: &str, _location: &log::LogLocation| {
		// This is a fairly simple format, though it's possible to do more complicated ones.
		// This closure can contain any code, as long as it produces a String message.
		format!("[{}][{}][{}] {}", time::now().strftime("%Y-%m-%d][%H:%M:%S").unwrap(), level, target, msg)
}),
```

"Logger implementations **typically** use the target to filter requests based on some user configuration", but logging it it gives more information about where the log event fired to the reader.

**Comment by giodamelio** [comment]
I am looking for the same functionality. It seems to me that passing in a [`&log::LogMetadata`](https://docs.rs/log/0.3.6/log/struct.LogMetadata.html) struct in place of the [`&log::LogeLevel`](https://docs.rs/log/0.3.6/log/enum.LogLevel.html) would be the best way to go about this. It would be a breaking change though. I am willing to make a pull request if it is ok with you @daboross.

**Comment by daboross** [comment]
I'm incredibly sorry I didn't reply to the last two comments when they were made. Thank you for the offer, and I'm always open to pull requests!

With that said, I've been working on an API rewrite for fern which includes this feature. It's almost a year since it was requested, but hopefully it's better late than never.

@giodamelio If you still have an interest in this library, could you possibly check out the changes I've committed to master, and see if you have any feedback on them?

I'm trying to make fern more ergonomic to use, and more runtime efficient.

The docs for the latest unreleased changes are at https://dabo.guru/rust/fern-dev/.

**Comment by daboross** [comment]
This is now included in `fern` 0.4!

Closing now that [`Dispatch::level_for`] exists.

[`Dispatch::level_for`]: https://dabo.guru/rust/fern/fern/struct.Dispatch.html#method.level_for
